snapcraft/internal/remote_build/_worktree.py

```diff
@@ -14,100 +14,263 @@
-            copy_function=self._link_and_add,
+        timestamp = datetime.datetime.now().strftime("%Y-%m-%d %H:%M")
+        self._git_repo.index.commit(
+            "snapcraft commit\n\nversion: {}\ntimestamp: {}\n".format(
+                snapcraft.__version__, timestamp
+            )
         )
 
-        git_dir = os.path.join(self._destroot, ".git")
-
-        # now remove files that exist only in destination
-        for dirpath, dirnames, filenames in os.walk(self._destroot, topdown=False):
-            if dirpath.startswith(git_dir):
-                continue
-            dirnames = [d for d in dirnames if d != ".git"]
-
-            for name in filenames:
-                dest = os.path.join(dirpath, name)
-                src = os.path.join(self._srcpath(dirpath), name)
-                if not os.path.exists(src):
-                    logger.debug("Remove file: {}".format(dest))
-                    self._unlink(dest)
-
-            for name in dirnames:
-                dest = os.path.join(dirpath, name)
-                src = os.path.join(self._srcpath(dirpath), name)
-                if not os.path.exists(src) and os.path.exists(dest):
-                    logger.debug("Remove directory: {}".format(dest))
-                    os.rmdir(dest)
-
-        # finally commit our changes, if any
-        self._repo.commit()
-
-    def add_remote(self, provider: str, user: str, build_id: str) -> str:
-        """Add a remote repository."""
-        return self._repo.add_remote(provider, user, build_id)
-
-    def push(self) -> None:
-        """Send our data to a remote repository."""
-        self._repo.push()
-
-    def _ignored_files(self, source: str, contents: List[str]) -> List[str]:
-        ignored = set()
-        for pattern in self._ignore:
-            ignored.update(set(fnmatch.filter(contents, pattern)))
-        return list(ignored)
-
-    def _link_and_add(self, source: str, destination: str) -> None:
-        link_or_copy(source, destination)
-        self._repo.add(os.path.relpath(destination, self._destroot))
-        logger.debug("Add file {}".format(destination))
-
-    def _unlink(self, path: str) -> None:
-        self._repo.remove(os.path.relpath(path, self._destroot))
-
-    def _srcpath(self, path: str) -> str:
-        rel = os.path.relpath(path, self._destroot)
-        return os.path.join(self._srcroot, rel)
+        return self._git_repo
+
+    def _process_part_sources(self, part_name: str, part_yaml: OrderedDict):
+        """Add part sources to tree.  Return source-modified yaml for part."""
+        logger.debug("processing {}: {}".format(part_name, part_yaml))
+
+        if "source" not in part_yaml:
+            logger.debug("no source specified, skipping...")
+            return part_yaml
+
+        source = part_yaml["source"]
+        source_modified = False
+        if isinstance(source, list):
+            # Grammar case.
+            part_yaml["source"] = []
+            for s in source:
```

Comment:
can you add a comment here about needing to take into account advanced grammar and maybe a snippet on how it looks?

Reply:
Oh you do a bit above...

Reply:
Yeah I'll add some detail for good measure, it is a bit of an unusual construct.